### PR TITLE
#161967125 add descriptive signup/registration validation errors.

### PR DESCRIPTION
**Pull request by jillo-abdullahi**
**What does this PR do?**

Update user registration validations and update validation errors.

**Description of tasks to be completed**

1. Add username validation for the following invalid inputs:

- Username length being less than 4 characters.
- Username having numbers only.

2. Add password validation for the following invalid inputs:
   

- Password length being less than 8.
- Password not containing at least one number.
- Password not containing at least one uppercase character.
- Password not containing at least one lowercase character.
- Password not containing at least one special character.

3. Add email validations error messages for duplicate email registration.
4. Update tests to cater for new validation error messages.

**How should this be manually tested?**

Clone the repo from your terminal: `git clone https://github.com/andela/ah-code-titans.git`

Create a virtual environment:`python3 -m venv <name_of_environment>`

Activate the virtual environment:  `source <name_of_environment>/bin/activate`

Install project dependencies: `pip install -r requirements.txt`

Add the following environment variables and populate them with your settings:

-`$ export DB_NAME='choose_a_dbname'`
-`$ export DB_USER='db_username'`
-`$ export DB_PASSWORD='db_role_password'`
-`$ export DB_HOST='localhost or preffered host'`
-`$ export DB_PORT='server_port_number'`
-`$ export SECRET_KEY='your_secret_key'`

Switch to the appropriate branch: `git checkout ft-validate-signup-161967125`
Run migrations: `python manage.py migrate`
Run tests: `python manage.py test`

**What are the relevant Pivotal Tracker stories?**
[#161967125](https://www.pivotaltracker.com/story/show/161967125)

**Checklist**

- [x] My code follows the style guidelines of this project

- [x] I have performed a self-review of my own code

- [x] I have commented my code, particularly in hard-to-understand areas

- [x] I have made corresponding changes to the documentation

- [x] My changes generate no new warnings

**Screenshots**

screenshot 2018-12-05 at 17 19 29

screenshot 2018-12-05 at 17 19 52

screenshot 2018-12-05 at 17 20 15

### CODE REVIEW

File: authors/settings.py [hunk 1]
```diff
@@ -157,14 +157,14 @@
...
-    'USE_SESSION_AUTH': False,
-    'DOC_EXPANSION': 'list',
-    'SECURITY_DEFINITIONS': {
-        'api_key': {
-            'type': 'apiKey',
-            'in': 'header',
-            'name': 'Authorization'
-        }
-    }
+  'SHOW_REQUEST_HEADERS': True,
```
Comment (Darius-Ndubi) [inline authors/settings.py]: This should not be changed, check indentation spaces

**Review by Darius-Ndubi — CHANGES_REQUESTED** [review]

File: authors/settings.py (hunk 1 above)
Comment (jillo-abdullahi) [reply authors/settings.py]: Thanks for noticing @Darius-Ndubi. I have updated this file.

**Review by Darius-Ndubi — APPROVED** [review]

**Review by Bryan-Cee — APPROVED** [review]
👍

**Review by davishooly — APPROVED** [review]

**Review by mutoromike — CHANGES_REQUESTED** [review]
Kindly add tests for the invalid user registration outcomes.

**Comment by mutoromike** [comment]
Also, fix the merge conflicts.

**Review by mutoromike — APPROVED** [review]

**Pull request merged**